src/app/page.tsx

```diff
@@ -87,6 +88,39 @@ export default async function Home(props: HomeProps) {
         itemFilters,
     } = parseParams(searchParams);
 
+    // Validate encounter exists if provided
+    if (encounter != null) {
+        const zones = await getZones();
+        const validEncounter = zones.some((zone) =>
+            zone.encounters.some((enc) => enc.id === encounter),
+        );
+
+        if (!validEncounter) {
+            notFound();
+        }
+    }
+
+    // Validate class exists if provided
+    if (classId != null) {
+        const classes = await getClasses();
+        const validClass = classes.some((cls) => cls.id === classId);
+
+        if (!validClass) {
+            notFound();
+        }
+    }
+
+    // Validate spec exists for the given class if both are provided
+    if (classId != null && specId != null) {
+        const classes = await getClasses();
+        const klass = classes.find((cls) => cls.id === classId);
+        const validSpec = klass?.specs.some((spec) => spec.id === specId);
+
+        if (!validSpec) {
+            notFound();
+        }
+    }
```

Comment:
If all three are provided, it means we will wait to fetch all three in series, right? can we not perform these checks in parallell to reduce any latency?

Reply:
Actually, I think we could probably handle the checking asyncrounously inside the `Promise.all()`, so which ever request succeeds first can be checked immidiately. 

So something like 

```
const promises = [];

if(enocunter) {
  promises.push(getZones().then(/* check validity in here, if invalid, call notFound() */))
}

if(classId) {
  // same thing
}

await Promise.all(promises)
```

That way we can fall back to 404 as soon as possible, but will not continue loading until both checks are done.